pkl-core/src/main/java/org/pkl/core/module/ProjectDependenciesManager.java

```diff
@@ -194,28 +206,42 @@ public Dependency getResolvedDependency(PackageUri packageUri) {
     return dep;
   }
 
-  public Path getProjectDir() {
-    return projectDir;
+  public URI getProjectBaseUri() {
+    return projectBaseUri;
   }
 
-  public Path getProjectDepsFile() {
-    return projectDir.resolve(PKL_PROJECT_DEPS_FILENAME);
+  public URI getProjectDepsFileUri() {
+    return projectBaseUri.resolve(PKL_PROJECT_DEPS_FILENAME);
+  }
+
+  public URI getProjectFileUri() {
+    return declaredDependencies.getProjectFileUri();
   }
 
   private ProjectDeps getProjectDeps() {
     synchronized (lock) {
       if (projectDeps == null) {
-        var depsPath = getProjectDepsFile();
-        if (!Files.exists(depsPath)) {
-          throw new VmExceptionBuilder().evalError("missingProjectDepsJson", projectDir).build();
-        }
+        var depsUri = getProjectDepsFileUri();
+        var moduleKey = moduleResolver.resolve(depsUri);
         try {
-          projectDeps = ProjectDeps.parse(depsPath);
-        } catch (IOException | URISyntaxException | JsonParseException e) {
+          // treat PklProject.deps.json as a module read, rather than introduce a new API.
+          var depsJson = moduleKey.resolve(securityManager).loadSource();
+          projectDeps = ProjectDeps.parse(depsJson);
+        } catch (IOException e) {
+          var exceptionBuilder =
+              new VmExceptionBuilder().evalError("cannotLoadProjectDepsJson", depsUri).withCause(e);
+          if (e.getMessage() != null) {
+            exceptionBuilder.withHint(e.getMessage());
+          } else {
+            exceptionBuilder.withHint("Encountered error: " + e);
+          }
+          throw exceptionBuilder.build();
```

Comment:
```suggestion
          throw new VmExceptionBuilder()
            .evalError("cannotLoadProjectDepsJson", depsUri)
            .withCause(e)
            .withHint(e.getMessage() != null ? e.getMessage() : ("Encountered error: " + e))
            .build();
```

Some day, null-coalescence will come to Java ;)